### ISSUE TITLE
Body sync failed

### ISSUE THREAD

**Issue by hashmap** [mimblewimble/grin#1845]
First it stopped syncing and just logging for 8 mins:
```
2018-10-25T16:22:49.804705157+02:00 DEBUG grin_servers::grin::sync::body_sync - body_sync: 0/0 blocks received in 5s
2018-10-25T16:22:49.804739265+02:00 DEBUG grin_servers::grin::sync::body_sync - body_sync: body_head - 027d28df, 5625, header_head - 0b12a183, 10474, sync_head - 0b12a183, 10474
2018-10-25T16:22:49.804759344+02:00 DEBUG grin_servers::grin::sync::body_sync - body_sync: cannot sync full blocks earlier than horizon.
```
after that the thread crashed:
```
2018-10-25T16:22:49.814945697+02:00 DEBUG grin_servers::common::types - sync_state: sync_status: BodySync { current_height: 5625, highest_height: 10475 } -> HeaderSync { current_height: 10474, highest_height: 10475 }
2018-10-25T16:22:49.814995924+02:00 DEBUG grin_servers::grin::sync::header_sync - sync: locator heights : [10474, 10472, 10468, 10460, 10444, 10412, 10348, 10220, 9964, 9452, 8428, 6380, 2284, 0]
2018-10-25T16:22:49.831846577+02:00 ERROR grin_util::logger -
thread 'sync' panicked at 'index out of bounds: the len is 0 but the index is 0': /checkout/src/libcore/slice/mod.rs:2085stack backtrace:
   0:     0x55c778a5333c - backtrace::backtrace::trace::h0d480c688c6e5983
   1:     0x55c778a530f2 - <backtrace::capture::Backtrace as core::default::Default>::default::hbf23cd498c62c039
   2:     0x55c778a53168 - backtrace::capture::Backtrace::new::hb00f34110e8fb8d2
   3:     0x55c7789e457e - grin_util::logger::send_panic_to_log::{{closure}}::h40696617360cc809
   4:     0x55c778b5fd73 - std::panicking::rust_panic_with_hook::h542dacac3f3a20d8
                        at libstd/panicking.rs:479
   5:     0x55c778b5f8d9 - std::panicking::continue_panic_fmt::hd989cd729c0c5a4b
                        at libstd/panicking.rs:390
   6:     0x55c778b5f7d5 - rust_begin_unwind
                        at libstd/panicking.rs:325
   7:     0x55c778bbaaeb - core::panicking::panic_fmt::h26b7cf0675909ac6
                        at libcore/panicking.rs:77
   8:     0x55c778bbaa9f - core::panicking::panic_bounds_check::hb7240398d20f0b17
                        at libcore/panicking.rs:59
   9:     0x55c7785ac4af - grin_servers::grin::sync::syncer::SyncRunner::sync_loop::h1b8b607d30e86ea8
  10:     0x55c77859a1cc - std::sys_common::backtrace::__rust_begin_short_backtrace::hf99d963c21c91c98
  11:     0x55c7785434bb - std::panicking::try::do_call::hbb2c5e4e2a879d08
  12:     0x55c778b84079 - __rust_maybe_catch_panic
                        at libpanic_unwind/lib.rs:105
  13:     0x55c7785591da - <F as alloc::boxed::FnBox<A>>::call_box::h1af53e7997867b92
  14:     0x55c778b622ea - <alloc::boxed::Box<(dyn alloc::boxed::FnBox<A, Output=R> + 'a)> as core::ops::function::FnOnce<A>>::call_once::h9a3d5ac4c96bb201
                        at /checkout/src/liballoc/boxed.rs:652
                         - std::sys_common::thread::start_thread::h24226b876494ff77
                        at libstd/sys_common/thread.rs:24
  15:     0x55c778b60285 - std::sys::unix::thread::Thread::new::thread_start::h0aab6afcd022cdd8
                        at libstd/sys/unix/thread.rs:90
  16:     0x7f5d963ada9c - start_thread
  17:     0x7f5d962c3a42 - clone
  18:                0x0 - <unknown>
```

**Comment by laboon** [comment]
I received the same (or a very similar) error when I tried to sync.  This was  on commit b889069722c3d99057e9cdd19138bc69a5126c84

Stack trace:
```
thread 'sync' panicked at 'index out of bounds: the len is 0 but the index is 0': libcore/slice/mod.rs:2046stack backtrace:
   0:        0x10baa6504 - backtrace::backtrace::trace::h38e3b1de9f341e04
                        at /Users/laboon/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.9/src/backtrace/mod.rs:42
   1:        0x10ba9f46c - backtrace::capture::Backtrace::new_unresolved::he95b8ee78352c3fe
                        at /Users/laboon/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.9/src/capture.rs:88
   2:        0x10ba9f3ce - backtrace::capture::Backtrace::new::h1e7d8b9d9ab04fb3
                        at /Users/laboon/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.9/src/capture.rs:63
   3:        0x10b92c2c7 - grin_util::logger::send_panic_to_log::{{closure}}::h8b29353009f12514
                        at /Users/laboon/pitt/grin/util/src/logger.rs:190
   4:        0x10bf495b8 - std::panicking::rust_panic_with_hook::h67318ff56366ae20
                        at /rustc/da5f414c2c0bfe5198934493f04c676e2b23ff2e/src/libstd/panicking.rs:481
   5:        0x10bf490cc - std::panicking::continue_panic_fmt::h2d9dbf1e069d262d
                        at /rustc/da5f414c2c0bfe5198934493f04c676e2b23ff2e/src/libstd/panicking.rs:391
   6:        0x10bf48fb8 - _rust_begin_unwind
                        at /rustc/da5f414c2c0bfe5198934493f04c676e2b23ff2e/src/libstd/panicking.rs:326
   7:        0x10bf8d8e1 - core::panicking::panic_fmt::h0f60ebf982c981f3
                        at /rustc/da5f414c2c0bfe5198934493f04c676e2b23ff2e/src/libcore/panicking.rs:77
   8:        0x10bf8d893 - core::panicking::panic_bounds_check::hcb523ab5839d27b7
                        at /rustc/da5f414c2c0bfe5198934493f04c676e2b23ff2e/src/libcore/panicking.rs:59
   9:        0x10acc8b08 - <usize as core::slice::SliceIndex<[T]>>::index::hf7913b411b85a974
                        at /Users/laboon/pitt/grin/libcore/slice/mod.rs:2046
  10:        0x10acdec24 - core::slice::<impl core::ops::index::Index<I> for [T]>::index::h929a0138673a8cf6
                        at /Users/laboon/pitt/grin/libcore/slice/mod.rs:1914
  11:        0x10acd8eba - <alloc::vec::Vec<T> as core::ops::index::Index<I>>::index::hdc7bd37ad24cfe76
                        at /Users/laboon/pitt/grin/liballoc/vec.rs:1725
  12:        0x10ac47359 - grin_servers::grin::sync::header_sync::HeaderSync::get_locator::h2e758f43d39f4bbc
                        at /Users/laboon/pitt/grin/servers/src/grin/sync/header_sync.rs:193
  13:        0x10ac46687 - grin_servers::grin::sync::header_sync::HeaderSync::request_headers::h033d931f8960a19a
                        at /Users/laboon/pitt/grin/servers/src/grin/sync/header_sync.rs:140
  14:        0x10ac46550 - grin_servers::grin::sync::header_sync::HeaderSync::header_sync::hb0b7434482d92409
                        at /Users/laboon/pitt/grin/servers/src/grin/sync/header_sync.rs:132
  15:        0x10ac46082 - grin_servers::grin::sync::header_sync::HeaderSync::check_run::h84ef37d3fb4d6ac8
                        at /Users/laboon/pitt/grin/servers/src/grin/sync/header_sync.rs:87
  16:        0x10ac5865f - grin_servers::grin::sync::syncer::SyncRunner::sync_loop::hcf42034d466427f8
                        at /Users/laboon/pitt/grin/servers/src/grin/sync/syncer.rs:148
  17:        0x10acbef15 - grin_servers::grin::sync::syncer::run_sync::{{closure}}::h66bf71385eff6d7c
                        at /Users/laboon/pitt/grin/servers/src/grin/sync/syncer.rs:38
  18:        0x10ac4575f - std::sys_common::backtrace::__rust_begin_short_backtrace::h97b1550f3c340493
                        at /Users/laboon/pitt/grin/libstd/sys_common/backtrace.rs:136
  19:        0x10acbbdbf - std::thread::Builder::spawn::{{closure}}::{{closure}}::he30ffb82095652cd
                        at /Users/laboon/pitt/grin/libstd/thread/mod.rs:409
  20:        0x10aca35af - <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::hf1cc03b90bc18fe8
                        at /Users/laboon/pitt/grin/libstd/panic.rs:313
  21:        0x10ac5781c - std::panicking::try::do_call::h22d51b1934940480
                        at /Users/laboon/pitt/grin/libstd/panicking.rs:310
  22:        0x10bf56dce - ___rust_maybe_catch_panic
                        at /rustc/da5f414c2c0bfe5198934493f04c676e2b23ff2e/src/libpanic_unwind/lib.rs:103
  23:        0x10ac572a4 - std::panicking::try::hbffb5ff9b43dd404
                        at /Users/laboon/pitt/grin/libstd/panicking.rs:289
  24:        0x10aca376f - std::panic::catch_unwind::h5264c433b7c3b892
                        at /Users/laboon/pitt/grin/libstd/panic.rs:392
  25:        0x10acbb6f0 - std::thread::Builder::spawn::{{closure}}::ha25b1d11ed0fd9ad
                        at /Users/laboon/pitt/grin/libstd/thread/mod.rs:408
  26:        0x10acbc227 - <F as alloc::boxed::FnBox<A>>::call_box::he7861a960b80e2ca
                        at /Users/laboon/pitt/grin/liballoc/boxed.rs:646
  27:        0x10bf48397 - std::sys_common::thread::start_thread::ha68f94e048f96d59
                        at /rustc/da5f414c2c0bfe5198934493f04c676e2b23ff2e/src/libstd/sys_common/thread.rs:24
  28:        0x10bf31018 - std::sys::unix::thread::Thread::new::thread_start::h57baa5b47ae41e12
                        at /rustc/da5f414c2c0bfe5198934493f04c676e2b23ff2e/src/libstd/sys/unix/thread.rs:90
  29:     0x7fff66d9333c - __pthread_body
  30:     0x7fff66d962a6 - __pthread_start
```

**Comment by laboon** [comment]
Looks like this is being worked on under issue 1834 - https://github.com/mimblewimble/grin/issues/1834

**Comment by Govhandled** [comment]
Gary yu

**Comment by antiochp** [comment]
I thought the horizon was `10,080` blocks (7 days).

Here we have - 

```
body_head - 027d28df, 5625, 
header_head - 0b12a183, 10474, 
sync_head - 0b12a183, 10474
```

So even if we have headers up to 10,474 we should be able to easily get blocks for `5,625`.

Edit: hang on - we are at `10,474` locally that doesn't mean our peers are there also...
Current chain head on the network is `27,937`.

So we need to be dropping back into state sync (for a new `txhashset.zip`) and we are not doing this here.

**Comment by hashmap** [comment]
@antiochp in latest log (debug info was added to catch this bug):
```
body_sync: cannot sync full blocks earlier than horizon. oldest_height: 0
```
so it tries to get block at height 0, function `is_on_current_chain` (or smth like that) is confused, probably by a missing header

**Comment by antiochp** [comment]
The underlying problem causing `NotFoundErr("BLOCK HEADER: 04064432")` should be resolved in #2009.
tl;dr - orphan block processing and header cache invalidation error.

Going to close this for now - we can reopen if it reoccurs.
